File: xforms/jvm/src/main/java/org/orbeon/oxf/xforms/processor/handlers/xhtml/XFormsSecretHandler.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.processor.handlers.xhtml;

import org.orbeon.xforms.XFormsConstants;
import org.orbeon.oxf.xforms.control.controls.XFormsSecretControl;
import org.orbeon.oxf.xml.XMLConstants;
import org.orbeon.oxf.xml.XMLReceiverHelper;
import org.orbeon.oxf.xml.XMLUtils;
import org.xml.sax.Attributes;
import org.xml.sax.ContentHandler;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.AttributesImpl;
import scala.Option;

/**
 * Handle xf:secret.
 */
public class XFormsSecretHandler extends XFormsControlLifecyleHandler {

    public XFormsSecretHandler(String uri, String localname, String qName, Attributes attributes, Object matched, Object handlerContext) {
        super(uri, localname, qName, attributes, matched, handlerContext, false, false);
    }

    @Override
    public void handleControlStart() throws SAXException {

        final XFormsSecretControl secretControl = (XFormsSecretControl) currentControl();
        final ContentHandler contentHandler = xformsHandlerContext.getController().getOutput();

        final AttributesImpl containerAttributes = getEmptyNestedControlAttributesMaybeWithId(getEffectiveId(), secretControl, true);

        // Create xhtml:input
        {
            final String xhtmlPrefix = xformsHandlerContext.findXHTMLPrefix();
            if (! isStaticReadonly(secretControl)) {
                final String inputQName = XMLUtils.buildQName(xhtmlPrefix, "input");
                containerAttributes.addAttribute("", "type", "type", XMLReceiverHelper.CDATA, "password");
                containerAttributes.addAttribute("", "name", "name", XMLReceiverHelper.CDATA, getEffectiveId());
                containerAttributes.addAttribute("", "value", "value", XMLReceiverHelper.CDATA,
                    secretControl == null || secretControl.getExternalValue() == null ? "" : secretControl.getExternalValue());

                // Handle accessibility attributes
                handleAccessibilityAttributes(attributes(), containerAttributes);
                handleAriaByAtts(containerAttributes);

                // Output all extension attributes
                // Output xxf:* extension attributes
                secretControl.addExtensionAttributesExceptClassAndAcceptForHandler(reusableAttributes, XFormsConstants.XXFORMS_NAMESPACE_URI());

                if (isXFormsReadonlyButNotStaticReadonly(secretControl))
                    outputReadonlyAttribute(reusableAttributes);

                handleAriaAttributes(secretControl.isRequired(), secretControl.isValid(), containerAttributes);

                // Output element
                contentHandler.startElement(XMLConstants.XHTML_NAMESPACE_URI(), "input", inputQName, containerAttributes);
                contentHandler.endElement(XMLConstants.XHTML_NAMESPACE_URI(), "input", inputQName);
            } else {
                // Output static read-only value
                final String spanQName = XMLUtils.buildQName(xhtmlPrefix, "span");
                containerAttributes.addAttribute("", "class", "class", "CDATA", "xforms-field");
                contentHandler.startElement(XMLConstants.XHTML_NAMESPACE_URI(), "span", spanQName, containerAttributes);

                final Option<String> value = secretControl.getFormattedValue();
                if (value.isDefined())
                    contentHandler.characters(value.get().toCharArray(), 0, value.get().length());

                contentHandler.endElement(XMLConstants.XHTML_NAMESPACE_URI(), "span", spanQName);
            }
        }
    }
}

File: xforms-runtime/shared/src/main/scala/org/orbeon/oxf/xforms/processor/handlers/xhtml/XFormsSecretHandler.scala
package org.orbeon.oxf.xforms.processor.handlers.xhtml

import org.orbeon.oxf.xforms.analysis.controls.{LHHAAnalysis, SecretControl}
import org.orbeon.oxf.xforms.control.controls.XFormsSecretControl
import org.orbeon.oxf.xforms.processor.handlers.XFormsBaseHandler.forwardAutocompleteAttribute
import org.orbeon.oxf.xforms.processor.handlers.{HandlerContext, XFormsBaseHandler}
import org.orbeon.oxf.xml.SaxSupport.*
import org.orbeon.oxf.xml.{XMLConstants, XMLUtils}
import org.orbeon.xforms.XFormsNames
import org.xml.sax.Attributes


class XFormsSecretHandler(
  uri            : String,
  localname      : String,
  qName          : String,
  attributes     : Attributes,
  elementAnalysis: SecretControl,
  handlerContext : HandlerContext
) extends
  XFormsControlLifecycleHandler(
    uri,
    localname,
    qName,
    attributes,
    elementAnalysis,
    handlerContext,
    repeating  = false,
    forwarding = false
  ) {

  private lazy val placeHolderInfo: Option[PlaceHolderInfo] =
    PlaceHolderInfo.placeHolderValueOpt(elementAnalysis, currentControl)

  def handleControlStart(): Unit = {

    val secretControl = currentControl.asInstanceOf[XFormsSecretControl]
    implicit val xmlReceiver = handlerContext.controller.output
    val containerAttributes = getEmptyNestedControlAttributesMaybeWithId(getEffectiveId, secretControl, addId = true)

    // Create `xh:input`
    val xhtmlPrefix = handlerContext.findXHTMLPrefix
    if (! XFormsBaseHandler.isStaticReadonly(secretControl)) {
      val inputQName = XMLUtils.buildQName(xhtmlPrefix, "input")
      containerAttributes.addOrReplace("type", "password")
      containerAttributes.addOrReplace("name", getEffectiveId)
      containerAttributes.addOrReplace("value", if (secretControl == null || secretControl.getExternalValue(handlerContext.collector) == null) "" else secretControl.getExternalValue(handlerContext.collector))
      // Handle accessibility attributes
      XFormsBaseHandler.forwardAccessibilityAttributes(attributes, containerAttributes, handlerContext)
      handleAriaByAtts(containerAttributes, XFormsLHHAHandler.coreControlLhhaByCondition)

      // Output all extension attributes
      // Output `xxf:*` extension attributes
      secretControl.addExtensionAttributesExceptClassAndAcceptForHandler(containerAttributes, XFormsNames.XXFORMS_NAMESPACE_URI)

      // Add attribute even if the control is not concrete
      placeHolderInfo foreach { placeHolderInfo =>
        if (placeHolderInfo.value ne null) // unclear whether this can ever be null
          containerAttributes.addOrReplace("placeholder", placeHolderInfo.value)
      }

      if (isXFormsReadonlyButNotStaticReadonly(secretControl))
        XFormsBaseHandlerXHTML.outputReadonlyAttribute(containerAttributes)
      XFormsBaseHandler.handleAriaAttributes(secretControl.isRequired, secretControl.isValid, secretControl.visited, containerAttributes)

      forwardAutocompleteAttribute(attributes, "input", containerAttributes)

      // Output element
      xmlReceiver.startElement(XMLConstants.XHTML_NAMESPACE_URI, "input", inputQName, containerAttributes)
      xmlReceiver.endElement(XMLConstants.XHTML_NAMESPACE_URI, "input", inputQName)
    } else {
      outputStaticReadonlyField(xhtmlPrefix) {
        secretControl.getFormattedValue(handlerContext.collector) foreach { value =>
          xmlReceiver.characters(value.toCharArray, 0, value.length)
        }
      }
    }
  }

  protected override def handleLabel(lhhaAnalysis: LHHAAnalysis): Unit =
    if (! (placeHolderInfo exists (_.isLabelPlaceholder)))
      super.handleLabel(lhhaAnalysis)

  protected override def handleHint(lhhaAnalysis: LHHAAnalysis): Unit =
    if (placeHolderInfo forall (_.isLabelPlaceholder))
      super.handleHint(lhhaAnalysis)
}